hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieGlobalBloomIndexV2.java

```diff
@@ -0,0 +1,223 @@
+        BloomFilter filter = fileIDToBloomFilter.get(partitionFileIdPair.getRight());
+        if (filter.mightContain(record.getRecordKey())) {
+          totalMatches++;
+          candidates.add(Pair.of(updatePartition(record, partitionFileIdPair.getLeft()), partitionFileIdPair.getRight()));
+        }
+      });
+      totalBloomCheckTimeMs += hoodieTimer.endTimer();
+
+      if (candidates.isEmpty()) {
+        candidates.add(Pair.of(record, ""));
+      }
+
+      totalCount++;
+      return candidates;
+    }
+
+    @Override
+    protected void start() {
+      totalTimeMs = 0;
+      totalMatches = 0;
+      totalCount = 0;
+      hoodieTimer = new HoodieTimer().startTimer();
+
+      // init global range bloomFilter
+      populateFileIDs();
+      populateRangeAndBloomFilters();
+      totalMetadataReadTimeMs += hoodieTimer.endTimer();
+      hoodieTimer.startTimer();
+    }
+
+    @Override
+    protected void end() {
+      totalTimeMs = hoodieTimer.endTimer();
+      String rangeCheckInfo = "LazyRangeAndBloomChecker: "
+              + "totalCount: " + totalCount + ", "
+              + "totalMatches: " + totalMatches + ", "
+              + "totalTimeMs: " + totalTimeMs + "ms, "
+              + "totalMetadataReadTimeMs: " + totalMetadataReadTimeMs + "ms, "
+              + "totalRangeCheckTimeMs: " + totalRangeCheckTimeMs + "ms, "
+              + "totalBloomCheckTimeMs: " + totalBloomCheckTimeMs + "ms";
+      LOG.info(rangeCheckInfo);
+    }
+
+    private HoodieRecord<T> updatePartition(HoodieRecord<T> record, String partitionPath) {
+      HoodieKey hoodieKey = new HoodieKey(record.getKey().getRecordKey(), partitionPath);
+      return new HoodieRecord<>(hoodieKey, record.getData());
+    }
+
+    private void populateFileIDs() {
+      try {
+        HoodieTableMetaClient metaClient = table.getMetaClient();
+        List<String> partitions = FSUtils.getAllPartitionPaths(metaClient.getFs(), metaClient.getBasePath(),
+                config.shouldAssumeDatePartitioning());
+
+        // Obtain the latest data files from all the partitions.
+        this.partitionPathFileIDList = partitions.stream().flatMap(partition -> {
+          Option<HoodieInstant> latestCommitTime =
+                  table.getMetaClient().getCommitsTimeline().filterCompletedInstants().lastInstant();
+          List<Pair<String, String>> filteredFiles = new ArrayList<>();
+          if (latestCommitTime.isPresent()) {
+            filteredFiles = table.getBaseFileOnlyView()
+                    .getLatestBaseFilesBeforeOrOn(partition, latestCommitTime.get().getTimestamp())
+                    .map(f -> Pair.of(partition, f.getFileId())).collect(Collectors.toList());
+          }
+          return filteredFiles.stream();
+        }).collect(Collectors.toList());
+      } catch (IOException e) {
+        throw new HoodieIOException("Failed to populateFileIDs", e);
+      }
+    }
+
+    private void populateRangeAndBloomFilters() {
```

Comment:
is there any difference between regular bloom IndexV2 and global version for this method? If most of the code base is similar, why can't we create a separate file for LazyRangeAndBloomChecker and introduce a global version(by extending from the same) and override only those required.